server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java

```diff
@@ -0,0 +1,533 @@
+     * @param parseContext  the ParseContext of the current document
+     */
+    public static void executePostParsePhases(MappingLookup lookup, ParseContext parseContext) throws IOException {
+        PostParsePhase postParsePhase = lookup.buildPostParsePhase(parseContext);
+        if (postParsePhase == null) {
+            return;
+        }
+        postParsePhase.execute();
+    }
+
+    PostParsePhase(
+        Map<String, PostParseExecutor> postParseExecutors,
+        Function<String, MappedFieldType> fieldTypeLookup,
+        ParseContext pc) {
+        LazyDocumentReader reader = new LazyDocumentReader(
+            pc.rootDoc(),
+            pc.sourceToParse().source(),
+            postParseExecutors.keySet());
+        this.context = new PostParseContext(fieldTypeLookup, pc, reader.getContext());
+        postParseExecutors.forEach((k, c) -> fieldExecutors.put(k, new OneTimeFieldExecutor(c)));
+    }
+
+    void execute() throws IOException {
+        for (OneTimeFieldExecutor executor : fieldExecutors.values()) {
+            executor.execute();
+        }
+    }
+
+    // FieldExecutors can be called both by executePostParse() and from the lazy reader,
+    // so to ensure that we don't run field scripts multiple times we guard them with
+    // this one-time executor class
+    private class OneTimeFieldExecutor {
+
+        PostParseExecutor executor;
+        boolean executed = false;
+
+        OneTimeFieldExecutor(PostParseExecutor executor) {
+            this.executor = executor;
+        }
+
+        void execute() throws IOException {
+            if (executed == false) {
+                try {
+                    executor.execute(context);
+                } catch (Exception e) {
+                    executor.onError(context, e);
+                }
+                executed = true;
+            }
+        }
+    }
+
+    private class LazyDocumentReader extends LeafReader {
+
+        private final ParseContext.Document document;
+        private final BytesReference sourceBytes;
+        private final Set<String> calculatedFields;
+        private final Set<String> fieldPath = new LinkedHashSet<>();
+
+        private LazyDocumentReader(ParseContext.Document document, BytesReference sourceBytes, Set<String> calculatedFields) {
+            this.document = document;
+            this.sourceBytes = sourceBytes;
+            this.calculatedFields = calculatedFields;
+        }
+
+        private void checkField(String field) throws IOException {
+            if (calculatedFields.contains(field)) {
+                // this means that a mapper script is referring to another calculated field;
+                // in which case we need to execute that field first. We also check for loops here
+                if (fieldPath.add(field) == false) {
+                    throw new IllegalStateException("Loop in field resolution detected: " + String.join("->", fieldPath) + "->" + field);
```

Comment:
I am not incredibly sure if this should be illegal state. It can effectively happen, so maybe it is more like an illegal argument exception?

Reply:
Oops, yes, that's me selecting the wrong exception type in autocomplete I think